crates/pera-core/src/signature_mpc/mpc_manager.rs

```diff
@@ -0,0 +1,450 @@
+    pub fn handle_mpc_events(&mut self, events: &Vec<Event>) -> anyhow::Result<()> {
+        if events.is_empty() {
+            return Ok(());
+        }
+        for event in events {
+            if P::InitEvent::type_() == event.type_ {
+                let deserialized_event = bcs::from_bytes(&event.contents)?;
+                self.push_new_mpc_instance(deserialized_event);
+                debug!("event: Init MPC Session {:?}", event);
+            };
+            if P::FinalizeEvent::type_() == event.type_ {
+                let deserialized_event = bcs::from_bytes(&event.contents)?;
+                self.finalize_mpc_instance(deserialized_event)?;
+                debug!("event: Finalize MPC Session {:?}", event);
+            };
+        }
+        Ok(())
+    }
+
+    /// Advance all the MPC instances that either received enough messages to, or perform the first step of the flow.
+    /// We parallelize the advances with Rayon to speed up the process.
+    pub async fn handle_end_of_delivery(&mut self) -> PeraResult {
+        let mut ready_to_advance = self
+            .mpc_instances
+            .iter_mut()
+            .filter_map(|(_, instance)| {
+                let threshold_number_of_parties = ((self.number_of_parties * 2) + 2) / 3;
+                if (instance.status == MPCSessionStatus::Active
+                    && instance.pending_messages.len() >= threshold_number_of_parties)
+                    || (instance.party.is_none() && instance.status == MPCSessionStatus::Active)
+                {
+                    Some(instance)
+                } else {
+                    None
+                }
+            })
+            .collect::<Vec<&mut SignatureMPCInstance<P>>>();
+        ready_to_advance
+            .par_iter_mut()
+            // TODO (#263): Mark and punish the malicious validators that caused some advances to return None, a.k.a to fail
+            .map(|ref mut instance| instance.advance(&self.auxiliary_input))
+            .collect::<PeraResult<_>>()?;
+        Ok(())
+    }
+
+    /// Handles a message by forwarding it to the relevant MPC instance
+    /// If the instance does not exist, punish the sender
+    pub fn handle_message(
+        &mut self,
+        message: &[u8],
+        authority_name: AuthorityName,
+        session_id: ObjectID,
+    ) -> PeraResult<()> {
+        let Some(mut instance) = self.mpc_instances.get_mut(&session_id) else {
+            error!(
+                "received a message for instance {:?} which does not exist",
+                session_id
+            );
+            // TODO (#261): Punish a validator that sends a message related to a non-existing mpc instance
+            return Ok(());
+        };
+        instance.handle_message(SignatureMPCMessage {
+            message: message.to_vec(),
+            authority: authority_name,
+        })
+    }
```

Comment:
we are processing the end of deliveries right now so it should not be an issue. Before start processing messages from delivery d+1, we finish process an instance at the end of delivery d

Reply:
@ItayLevyOfficial true, but this will change when [#282](https://github.com/dwallet-labs/dwallet-network/issues/282) is
addressed, so let's remember this here.